src/java/voldemort/utils/pool/KeyedResourcePool.java

```diff
@@ -123,35 +126,53 @@ public V checkout(K key) throws Exception {
         checkNotClosed();
 
         long startNs = System.nanoTime();
+        long timeoutNs = resourcePoolConfig.getTimeout(TimeUnit.NANOSECONDS);
+        long endNs = startNs + timeoutNs;
         Pool<V> resourcePool = getResourcePoolForKey(key);
         V resource = null;
         try {
             checkNotClosed();
-            // Must attempt a non blocking checkout before blockingGet to ensure
-            // resources are created for the pool.
-            resource = attemptNonBlockingCheckout(key, resourcePool);
+            long totalNonBlockingElapsedNs = 0;
+            long iterStartTime = 0;
+            long totalBlockingElapsedNs = 0;
+            final long MAX_WAIT_TIME = 300 * Time.NS_PER_MS;
 
-            if(resource == null) {
-                long nonBlockingElapsedNs = System.nanoTime() - startNs;
-                long timeRemainingNs = resourcePoolConfig.getTimeout(TimeUnit.NANOSECONDS)
-                                       - nonBlockingElapsedNs;
+            while(resource == null && (iterStartTime = System.nanoTime()) < endNs) {
```

Comment:
Looks like you spin until you get a connection checked out? Won't you lock up the selector manager threads.. (Thats one of the reasons dr.j went this route)